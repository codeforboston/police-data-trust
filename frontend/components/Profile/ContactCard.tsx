"use client"

import { Card, CardContent, Typography, IconButton } from "@mui/material"
import ModeEditOutlinedIcon from "@mui/icons-material/ModeEditOutlined"
import EmailIcon from "@mui/icons-material/Email"
import PublicIcon from "@mui/icons-material/Public"
import FacebookIcon from "@mui/icons-material/Facebook"
import InstagramIcon from "@mui/icons-material/Instagram"
import LinkedInIcon from "@mui/icons-material/LinkedIn"
import YouTubeIcon from "@mui/icons-material/YouTube"
import XIcon from "@mui/icons-material/X"
import { useRouter } from "next/navigation"
import styles from "./contactCard.module.css"

interface Props {
  primaryEmail: string
  secondaryEmail?: string
  website?: string
  socials: {
    facebook?: string
    instagram?: string
    linkedin?: string
    twitter?: string
    youtube?: string
  }
}

export default function ContactCard({ primaryEmail, secondaryEmail, website, socials }: Props) {
  const router = useRouter()
  const hasSocials = Object.values(socials).some((val) => !!val)

  // normalize website URLs to always include http(s)
  const getSafeUrl = (url: string) => {
    if (!url.startsWith("http://") && !url.startsWith("https://")) {
      return `https://${url}`
    }
    return url
  }

  return (
    <Card variant="outlined" sx={{ marginTop: "20px", marginBottom: "40px" }}>
      <CardContent
        sx={{
          p: "40px",
          "&:last-child": { pb: "40px" },
          "@media (max-width:430px)": {
            p: "24px",
            "&:last-child": { pb: "24px" }
          }
        }}>
        <IconButton className={styles.editIcon} sx={{ color: "#000" }}>
          <ModeEditOutlinedIcon onClick={() => router.push("/profile/contact/edit")} />
        </IconButton>

        <Typography variant="h5" fontWeight={600}>
          Contact
        </Typography>

        <div className={styles.contactWrapper}>
          <div className={styles.contactRow}>
            <EmailIcon />
            <div className={styles.contactDetails}>
              <p className={styles.contactLabel}>Email</p>
              <a href={`mailto:${primaryEmail}`} className={styles.emailLink}>
                {primaryEmail}
              </a>
            </div>
          </div>

          {secondaryEmail && (
            <div className={styles.contactRow}>
              <EmailIcon />
              <div className={styles.contactDetails}>
                <p className={styles.contactLabel}>Secondary Email</p>
                <a href={`mailto:${secondaryEmail}`} className={styles.emailLink}>
                  {secondaryEmail}
                </a>
              </div>
            </div>
          )}

          {website && (
            <div className={styles.contactRow}>
              <PublicIcon />
              <div className={styles.contactDetails}>
                <p className={styles.contactLabel}>Website</p>
                <a
                  href={getSafeUrl(website)}
                  target="_blank"
                  rel="noopener noreferrer"
                  className={styles.websiteLink}>
                  {website}
                </a>
              </div>
            </div>
          )}
        </div>

        {hasSocials && (
          <>
            <Typography variant="h5" fontWeight={600} sx={{ marginTop: "24px" }}>
              Socials
            </Typography>
            <div className={styles.socials}>
              {socials.linkedin && (
                <a href={socials.linkedin} target="_blank" rel="noopener noreferrer">
                  <LinkedInIcon fontSize="large" />
                </a>
              )}
              {socials.facebook && (
                <a href={socials.facebook} target="_blank" rel="noopener noreferrer">
                  <FacebookIcon fontSize="large" />
                </a>
              )}
              {socials.instagram && (
                <a href={socials.instagram} target="_blank" rel="noopener noreferrer">
                  <InstagramIcon fontSize="large" />
                </a>
              )}
              {socials.twitter && (
                <a href={socials.twitter} target="_blank" rel="noopener noreferrer">
                  <XIcon fontSize="large" />
                </a>
              )}
              {socials.youtube && (
                <a href={socials.youtube} target="_blank" rel="noopener noreferrer">
                  <YouTubeIcon fontSize="large" />
                </a>
              )}
            </div>
          </>
        )}
      </CardContent>
    </Card>
  )
}
