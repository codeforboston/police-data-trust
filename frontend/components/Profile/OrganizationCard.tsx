import React from "react"
import { Avatar, Card, CardContent, Typography } from "@mui/material"
import styles from "./organizationCard.module.css"

export default function OrganizationCard() {
  return (
    <Card variant="outlined" sx={{ marginTop: "20px" }}>
      <CardContent
        sx={{
          p: "40px",
          "&:last-child": {
            pb: "40px"
          },
          "@media (max-width:430px)": {
            p: "24px",
            "&:last-child": {
              pb: "24px"
            }
          }
        }}
      >
        <Typography variant="h5" fontWeight={600}>
          Organization Affiliation
        </Typography>
        <div className={styles.container}>
          <Avatar sx={{ width: 40, height: 40 }} src={"/broken-image.jpg"} />
          <div className={styles.containerText}>
            <Typography fontWeight={500} fontSize={20}>
              Organization Name
            </Typography>
            <Typography>Bio from the organization’s profile.</Typography>
          </div>
        </div>
        <Typography color="text.secondary">Joined on Oct 18, 2024</Typography>
        <div className={styles.container}>
          <Avatar sx={{ width: 40, height: 40 }} src={"/broken-image.jpg"} />
          <div className={styles.containerText}>
            <Typography fontWeight={500} fontSize={20}>
              Organization Name
            </Typography>
            <Typography>Bio from the organization’s profile.</Typography>
          </div>
        </div>
        <Typography color="text.secondary">Joined on Oct 18, 2024</Typography>
      </CardContent>
    </Card>
  )
}
