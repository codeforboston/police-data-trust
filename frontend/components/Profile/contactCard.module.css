.editIcon {
  float: right;
  cursor: pointer;
}

.contactWrapper {
  display: flex;
  flex-direction: column;
  gap: 16px;
  margin-top: 16px;
}

.contactRow {
  display: flex;
  gap: 12px;
}

.contactDetails {
  display: flex;
  flex-direction: column;
}

.contactLabel {
  font-weight: 500;
  font-size: 20px;
}

.socials {
  display: flex;
  gap: 16px;
  margin-top: 8px;
}

.emailLink,
.websiteLink {
  text-decoration: none;
  word-break: break-all;
}

.emailLink:hover,
.websiteLink:hover {
  text-decoration: underline;
}
