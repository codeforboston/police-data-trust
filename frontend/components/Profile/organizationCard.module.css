.container {
  display: flex;
  gap: 12px;
  margin-top: 24px;
  align-items: center;
}

.containerText {
  display: flex;
  flex-direction: column;
}
