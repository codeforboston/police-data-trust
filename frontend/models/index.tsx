export * from "./app-routes"
export * from "./enrollment-cta"
export * from "./info-tooltip"
export * from "./logo-sizes"
export * from "./password-aid"
export * from "./primary-input"
export * from "./secondary-input"
export * from "./response"
export * from "./results-alert"
export * from "./saved-table"
export * from "./state-select"
export * from "./toggle-options"
export * from "./visualizations"
