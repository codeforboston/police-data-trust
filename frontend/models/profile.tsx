// User Profile models
import { User } from "../helpers/api"

export enum ProfileMenu {
  USER_INFO = "info",
  PROFILE_TYPE = "type",
  SAVED_SEARCHES = "searches",
  SAVED_RESULTS = "results",
  NOTIFICATIONS = "notifications",
  ORGANIZATIONS = "organizations"
}

interface MenuText {
  item: ProfileMenu
  text: string
}

export const profileMenuItems: MenuText[] = [
  { item: ProfileMenu.USER_INFO, text: "User Information" },
  { item: ProfileMenu.PROFILE_TYPE, text: "Profile Type" },
  { item: ProfileMenu.SAVED_RESULTS, text: "Saved Results" },
  { item: ProfileMenu.SAVED_SEARCHES, text: "Saved Searches" },
  { item: ProfileMenu.NOTIFICATIONS, text: "Notifications" },
  { item: ProfileMenu.ORGANIZATIONS, text: "My Organizations" }
]

// props for profile menu content
export interface UserProfileProps {
  userData: UserDataType
}

/**
 * from backend/database/models/user.py
 */
export enum UserRoles {
  NONE = 0,
  PASSPORT = 1,
  PUBLIC = 2,
  CONTRIBUTOR = 3,
  ADMIN = 4
}

// User Information
export interface UserDataType {
  active: boolean
  firstName: string
  lastName: string
  email: string
  phone: string
  role: UserRoles
}

export const emptyUser: UserDataType = {
  active: false,
  firstName: "",
  lastName: "",
  email: "",
  phone: "",
  role: 0
}

export const publicUser = (user: User): UserDataType => ({
  firstName: user.firstName || "",
  lastName: user.lastName || "",
  email: user.email,
  phone: user.phoneNumber,
  active: user.active,
  role: UserRoles.PUBLIC
})

export const passportUser = (user: User): UserDataType => ({
  firstName: user.firstName || "",
  lastName: user.lastName || "",
  email: user.email,
  phone: user.phoneNumber,
  active: user.active,
  role: UserRoles.PASSPORT
})

export const contributorUser = (user: User): UserDataType => ({
  firstName: user.firstName || "",
  lastName: user.lastName || "",
  email: user.email,
  phone: user.phoneNumber,
  active: user.active,
  role: UserRoles.CONTRIBUTOR
})

export const adminUser = (user: User): UserDataType => ({
  firstName: user.firstName || "",
  lastName: user.lastName || "",
  email: user.email,
  phone: user.phoneNumber,
  active: user.active,
  role: UserRoles.ADMIN
})

export const someUser = (user: User, role: UserRoles): UserDataType => ({
  firstName: user.firstName || "",
  lastName: user.lastName || "",
  email: user.email,
  phone: user.phoneNumber,
  active: user.active,
  role: role
})

// Account Type
interface ProfileTypeText {
  title: string
  content: string
  linkText?: string
  linkPath?: string
}

export const profileTypeContent: { [key in UserRoles]: ProfileTypeText } = {
  [UserRoles.NONE]: {
    title: "No Account",
    content: "",
    linkText: "Sign up for an account",
    linkPath: "/register"
  },
  [UserRoles.PUBLIC]: {
    title: "Viewer Account",
    content:
      "A Viewer account allows you to see data that is publicly available. To apply to see legally protected data...",
    linkText: "Register for a Passport Account",
    linkPath: "/passport"
  },
  [UserRoles.PASSPORT]: {
    title: "Passport Account",
    content:
      "A Passport account allows you to see legally protected data. If the publicly available data is sufficient for your needs...",
    linkText: "Revert to a Viewer Account",
    linkPath: "/register"
  },
  [UserRoles.CONTRIBUTOR]: {
    title: "Contributor Account",
    content: ""
  },
  [UserRoles.ADMIN]: {
    title: "Administrator Account",
    content: ""
  }
}
