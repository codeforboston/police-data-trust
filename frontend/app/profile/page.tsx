"use client"

import React from "react"
import { useUserProfile } from "@/utils/useUserProfile"
import SuggestionsCard from "@/components/Profile/SuggestionsCard"
import ProfileLayout from "@/components/Profile/ProfileLayout"
import ProfileHeaderCard from "@/components/Profile/ProfileHeaderCard"
import OrganizationCard from "@/components/Profile/OrganizationCard"
import ContactCard from "@/components/Profile/ContactCard"
import { SocialMedia } from "@/utils/api"

export default function ProfilePage() {
  const { profile, loading } = useUserProfile()

  if (loading) return <p>Loading profile...</p>
  if (!profile) return <p>Unable to load profile.</p>

  const peopleSuggestions = [
    { name: "Samuel Smith", title: "Title", avatarUrl: "/broken-image.jpg" },
    { name: "Marian Linehan", title: "Title", avatarUrl: "/broken-image.jpg" },
    { name: "June MacCabe", title: "Title", avatarUrl: "/broken-image.jpg" },
    { name: "Joseph Vanasse", title: "Title", avatarUrl: "/broken-image.jpg" }
  ]

  const orgSuggestions = [
    { name: "Law Firm Name 1", title: "Title", avatarUrl: "/broken-image.jpg" },
    { name: "Law Firm Name 2", title: "Title", avatarUrl: "/broken-image.jpg" },
    { name: "Law Firm Name 3", title: "Title", avatarUrl: "/broken-image.jpg" },
    { name: "Law Firm Name 4", title: "Title", avatarUrl: "/broken-image.jpg" }
  ]

  const socialMediaContacts = profile.social_media || ({} as SocialMedia)

  return (
    <ProfileLayout
      sidebar={
        <>
          <SuggestionsCard title="People you may know" items={peopleSuggestions} />
          <SuggestionsCard title="Organizations you may know" items={orgSuggestions} />
        </>
      }>
      <ProfileHeaderCard
        firstName={profile.first_name}
        lastName={profile.last_name}
        avatarUrl={profile.profile_image}
        title={profile.employment?.title || ""}
        organization={profile.employment?.employer || ""}
        city={profile.location?.city || ""}
        state={profile.location?.state || ""}
        biography={profile.bio || ""}
        isOwnProfile={true}
      />
      <OrganizationCard />
      <ContactCard
        primaryEmail={profile.primary_email || ""}
        secondaryEmail={profile.contact_info.additional_emails[0]}
        website={profile.website || undefined}
        socials={{
          facebook: socialMediaContacts.facebook_url || undefined,
          instagram: socialMediaContacts.instagram_url || undefined,
          linkedin: socialMediaContacts.linkedin_url || undefined,
          twitter: socialMediaContacts.twitter_url || undefined,
          youtube: socialMediaContacts.youtube_url || undefined
        }}
      />
    </ProfileLayout>
  )
}
