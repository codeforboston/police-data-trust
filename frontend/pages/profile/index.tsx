import * as React from "react"
import {
  ProfileInfo,
  ProfileNav,
  ProfileNotifications,
  ProfileOrganizations,
  ProfileType,
  SavedResults,
  SavedSearches
} from "../../compositions"
import { requireAuth } from "../../helpers"
import { ProfileMenu } from "../../models/profile"
import { Layout } from "../../shared-components"
import styles from "./profile.module.css"
import OrgUserTable from "../../compositions/profile-orguser/profile-orguser"

export default requireAuth(function Profile() {
  const [activePage, setActivePage] = React.useState(ProfileMenu.USER_INFO)

  const ActivePageComp = (function (menuItem: ProfileMenu) {
    switch (menuItem) {
      case ProfileMenu.USER_INFO:
        return ProfileInfo
      case ProfileMenu.PROFILE_TYPE:
        return ProfileType
      case ProfileMenu.SAVED_RESULTS:
        return SavedResults
      case ProfileMenu.SAVED_SEARCHES:
        return SavedSearches
      case ProfileMenu.NOTIFICATIONS:
        return ProfileNotifications
      case ProfileMenu.ORGANIZATIONS:
        // return ProfileOrganizations
        return OrgUserTable
      default:
        throw new Error("Must be a key in 'ProfileMenu' enum - unexpected default case!")
    }
  })(activePage)

  return (
    <Layout>
      <div className={styles.profileWrapper}>
        <ProfileNav activePage={activePage} setActivePage={setActivePage} />
        <ActivePageComp />
      </div>
    </Layout>
  )
})
