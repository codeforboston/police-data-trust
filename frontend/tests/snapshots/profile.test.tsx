import { ProfileMenu } from "../../models/profile"
import Profile from "../../pages/profile"
import { render, setAuthForTest } from "../test-utils"

beforeAll(() => setAuthForTest())

it("renders Profile Page correctly", () => {
  const { container } = render(<Profile initialTab={ProfileMenu.USER_INFO} />)
  expect(container).toMatchSnapshot()
})
