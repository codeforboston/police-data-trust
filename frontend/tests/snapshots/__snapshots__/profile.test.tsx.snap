// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Profile Page correctly 1`] = `
<div>
  <div
    style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
  />
  <div>
    <div
      class="profileWrapper"
    >
      <nav
        class="leftMenu"
      >
        <ul>
          <li
            class="selectedItem"
          >
            User Information
          </li>
          <li
            class="menuItem"
          >
            Profile Type
          </li>
          <li
            class="menuItem"
          >
            Saved Results
          </li>
          <li
            class="menuItem"
          >
            Saved Searches
          </li>
          <li
            class="menuItem"
          >
            Notifications
          </li>
          <li
            class="menuItem"
          >
            My Organizations
          </li>
        </ul>
      </nav>
      <div
        class="profileData"
      >
        <header
          class="sectionTitle"
        >
          Your Account Information
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-edit fa-w-18 fa-1x editButton"
            data-icon="edit"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 576 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M402.6 83.2l90.2 90.2c3.8 3.8 3.8 10 0 13.8L274.4 405.6l-92.8 10.3c-12.4 1.4-22.9-9.1-21.5-21.5l10.3-92.8L388.8 83.2c3.8-3.8 10-3.8 13.8 0zm162-22.9l-48.8-48.8c-15.2-15.2-39.9-15.2-55.2 0l-35.4 35.4c-3.8 3.8-3.8 10 0 13.8l90.2 90.2c3.8 3.8 10 3.8 13.8 0l35.4-35.4c15.2-15.3 15.2-40 0-55.2zM384 346.2V448H64V128h229.8c3.2 0 6.2-1.3 8.5-3.5l40-40c7.6-7.6 2.2-20.5-8.5-20.5H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V306.2c0-10.7-12.9-16-20.5-8.5l-40 40c-2.2 2.3-3.5 5.3-3.5 8.5z"
              fill="currentColor"
            />
          </svg>
        </header>
        <main
          class="dataView"
        >
          <div
            class="dataCell"
          >
            <div
              class="label"
            >
              First name:
            </div>
            <div
              class="dataField"
            >
              FirstTest
            </div>
          </div>
          <div
            class="dataCell"
          >
            <div
              class="label"
            >
              Last name:
            </div>
            <div
              class="dataField"
            >
              LastTest
            </div>
          </div>
          <div
            class="dataBreak"
          />
          <div
            class="dataCell"
          >
            <div
              class="label"
            >
              Email address:
            </div>
            <div
              class="dataField"
            >
              testemail@example.com
            </div>
          </div>
          <div
            class="dataCell"
          >
            <div
              class="label"
            >
              Phone number:
            </div>
            <div
              class="dataField"
            >
              --
            </div>
          </div>
        </main>
      </div>
    </div>
  </div>
</div>
`;
